addon/components/frost-file-picker.js

```diff
@@ -20,40 +18,66 @@ export default Component.extend({
       }
     })
   }),
-
   bindChange: on('didInsertElement', function () {
     this.$('.frost-file-select').on('change', run.bind(this, 'filesSelected'))
   }),
-
   unbindChange: on('willDestroyElement', function () {
     this.$('.frost-file-select').off('change', run.bind(this, 'filesSelected'))
   }),
-
-  filesSelected: function (event) {
-    this.set('fileName', null)
-
-    let files = event.target.files
-
-    if (typeof (this.attrs['validate']) === 'function') {
-      this.attrs['validate'](files[0]).then((result) => {
-        if (result.valid) {
-          this.set('fileName', files[0].name)
-          if (_.isFunction(this.attrs['onChange'])) {
-            this.attrs['onChange']({id: this.get('id'), type: 'file', value: files[0]})
-          }
-        }
+  fireHandler: function (file) {
+    if (typeof this.attrs['onChange'] === 'function') {
+      this.attrs['onChange']({
+        id: this.get('id'),
+        type: 'file',
+        value: file
       })
-    } else {
-      this.set('fileName', files[0].name)
-      if (_.isFunction(this.attrs['onChange'])) {
-        this.attrs['onChange']({id: this.get('id'), type: 'file', value: files[0]})
-      }
     }
   },
-
+  filesSelected: function (files) {
+    if (!(files instanceof window.FileList)) {
+      files = files.target.files
+    }
+    files = [].slice.call(files)
+    let self = this
+    let name = files.reduce((e, r) => `${e}${r.name}, `, '')
+    this.set('fileName', name.slice(0, -2))
+    files.forEach((file) => {
+      new Promise(function (resolve, reject) {
+        if (typeof (self.attrs['validate']) === 'function') {
+          self.attrs['validate'](file).then((result) => {
+            result.valid ? resolve(file) : reject(result)
+          })
+        } else {
+          try {
+            resolve(file)
+          } catch (e) {
+            reject(e)
+          }
+        }
+      }).then(this.get('fireHandler').bind(this))
+        .catch(this.attrs['onError'])
+    })
+  },
```

Comment:
This function should be split into smaller functions to improve readability. A good start would be:
```js
_getFiles (files) {
  if (!(files instanceof window.FileList)) {
    files = files.target.files
  }

  return [].slice.call(files)
},

_updateFileName (files) {
  const name = files.reduce((e, r) => `${e}${r.name}, `, '')
  this.set('fileName', name.slice(0, -2))
},

filesSelected (files) {
  files = getFiles(files)
  this._updateFileName(files)

  files.forEach((file) => {
    new Promise((resolve, reject) => {
      if (typeof (this.attrs['validate']) === 'function') {
        this.attrs['validate'](file).then((result) => {
          result.valid ? resolve(file) : reject(result)
        })
      } else {
        try {
          resolve(file)
        } catch (e) {
          reject(e)
        }
      }
    })
      .then(this.get('fireHandler').bind(this))
      .catch(this.attrs['onError'])
  })
},
```